Adding command line interface to model

ericbarefoot opened the pull request:
### 📣 New Feature 📣 
Per thoughts from @amoodie, I'm adding here a set of interfaces into the pyDeltaRCM model that allow users to interface through the command line. 

The two new basic ways to do this are 
(1) through a command-line entry point and 
(2) through `__main__.py`. Both are demo-ed below.
```bash
# interface 1
run_pyDeltaRCM                                  # runs the model using default values
run_pyDeltaRCM --version                        # returns the version
run_pyDeltaRCM --config <path_to_config_here>   # runs the model using a config file
```
This method works by establishing a `console_scripts` entry point in the `setup.py` file, so when the user `pip install`s, they get a command line hook. (see [this line](https://github.com/ericbarefoot/pyDeltaRCM_WMT/blob/b5a19c22f6df3493b01200805bd98f0d237195de/setup.py#L19))
```bash
# interface 2
python -m pyDeltaRCM                                  # runs the model using default values
python -m pyDeltaRCM --version                        # returns the version
python -m pyDeltaRCM --config <path_to_config_here>   # runs the model using a config file
```
This method works by simply calling the same command line script function in a `__main__.py` file, so the two are essentially just aliases for each other. 

#### Other new things

In addition to the main feature, to make this work as a default, I [included a new `timesteps` parameter in the `default.yml`](https://github.com/ericbarefoot/pyDeltaRCM_WMT/blob/b5a19c22f6df3493b01200805bd98f0d237195de/pyDeltaRCM/default.yml#L145-L147), which lets the user specify timesteps for the simplest case of running the model with no special stuff. 

I also added a new `test.yml` because I'm not really sure how to test stuff from within the command line, so I figured a good way would be to output a file, and check that file. In the future, we can think about having this file be data output, and if necessary, load it and a saved version and check for congruence. 

This PR mostly fixes #22, but doesn't include the last component, which is the matrix expansion of defaults. I suggest we postpone that, and make it a new issue.

ericbarefoot committed:
feat: add command line hook for deltaRCM

I agree with Andrew, that the command line would be a good way to 
interface with the model, especially if configuration files could be 
passed from the command line.

In this commit, I add that feature using a console_scripts entrypoint in 
setup.py to a special new function called run_model, and this function 
runs the model for a given number of timesteps. This number of timesteps 
is set by the config file, which now accepts an additional timesteps 
parameter.

ericbarefoot committed:
tests: add unit tests for new command line function

Basically, since we don't get a model instance to work with when we test 
from command line, I just hacked this together to save an output image, 
and check if that image exists. This is enabled with a new 
test_output.yaml file.

I also added a test to make sure the versions match when we query from 
command line.

ericbarefoot committed:
feat: add __main__.py for execution from command line python

now execution from a config file can be initiated either from a bash 
entrypoint or from a main file when called via 'python -m pyDeltaRCM' 
This change addresses #22.

ericbarefoot committed:
tests: add unit tests for python hook into model.

ericbarefoot committed:
fix: make test for command line compatible with python 3.6

Issue was with the subprocess package. I was using kw arguments that 
were introduced in 3.7, so I reverted the call to the 3.6 version, now 
it should run on both.

amoodie commented on pyDeltaRCM/command_line.py:
```diff
@@ -0,0 +1,34 @@
+import os
+import argparse
+
+import pyDeltaRCM as deltaModule
```
why? can't you get pyDeltaRCM version from `v = pydeltarcm.shared_tools._get_version()` ([here](https://github.com/DeltaRCM/pyDeltaRCM_WMT/blob/fd0be0d091af7056a4004cace13c476ca5163b3d/pyDeltaRCM/shared_tools.py#L41)) or, even better, from `pyDeltaRCM.__version__` ([here](https://github.com/DeltaRCM/pyDeltaRCM_WMT/blob/fd0be0d091af7056a4004cace13c476ca5163b3d/pyDeltaRCM/__init__.py#L7)) I think this would be preferred to declaring an import convention we don't use anywhere else.

ericbarefoot replied:
Agreed. as you mention in your next comment, the issue was that the namespaces conflicted, and this was the solution I came up with while messing around. I'll do what you say below I think.

amoodie commented on pyDeltaRCM/command_line.py:
```diff
@@ -0,0 +1,34 @@
+import argparse
+
+import pyDeltaRCM as deltaModule
+from pyDeltaRCM.deltaRCM_driver import pyDeltaRCM
```
how about `from . import deltaRCM_driver` and then below use `deltaRCM_driver.pyDeltaRCM`. I think this is clearer as to which namespace you are using.  

It also confirms why you are probably right we should change the name of the "model" class....

ericbarefoot replied:
I like this solution.
>It also confirms why you are probably right we should change the name of the "model" class....

Yeah. This was what I was kindof getting at. It just means that if you want to access the module, you have to do screwy namespace stuff to avoid conflicts.

amoodie commented on pyDeltaRCM/command_line.py:
```diff
@@ -0,0 +1,34 @@
+
+    for time in range(0, delta.timesteps):
+
+        print(delta._time)
```
I think this is handled [here](https://github.com/DeltaRCM/pyDeltaRCM_WMT/blob/fd0be0d091af7056a4004cace13c476ca5163b3d/pyDeltaRCM/deltaRCM_tools.py#L36) by the verbosity setting, so if verbosity were set to `>0` you would get double printing. 

Better just leave it where it's already implemented I think?

ericbarefoot replied:
Yep. good catch. I suspect I turned it on b/c this laptop is slow, and it showed me that progress was happening. And I think test.yaml has 0 verbosity.

amoodie commented on pyDeltaRCM/command_line.py:
```diff
@@ -0,0 +1,34 @@
+        print(delta._time)
+        delta.update()
+
+        delta.finalize()
```
shouldn't this happen only once, and after the loop? or do I not understand what `delta.finalize()` does

ericbarefoot replied:
You're correct. Also a good catch.

amoodie commented on setup.py:
```diff
@@ -2,6 +2,7 @@
 from setuptools import setup, find_packages
 
 from pyDeltaRCM import shared_tools
+import pyDeltaRCM
```
Is it possible to call the `command_line:run_model` more directly than total package import? I'm afraid this line would fail for any number of reasons totally unrelated to the command line runner and it could be annoying to debug.

ericbarefoot replied:
I changed this to a specific import of `from pyDeltaRCM import command_line`.

amoodie reviewed (commented):
Looks neat @ericbarefoot! I made a few suggestions below, mostly about imports, since this is mostly importing code lol. 

I think the api is top notch, if we decide to do a different hook down the line (as discussed), I think the api would be unchanged, so I'm good with all the high level stuff.

amoodie commented:
also forgot to comment on:

> I also added a new test.yml because I'm not really sure how to test stuff from within the command line, so I figured a good way would be to output a file, and check that file. In the future, we can think about having this file be data output, and if necessary, load it and a saved version and check for congruence.

Yeah, tests look totally :ok_hand: to me!

> This PR mostly fixes #22, but doesn't include the last component, which is the matrix expansion of defaults. I suggest we postpone that, and make it a new issue.

Yeppers, sounds good.

elbeejay reviewed (commented):
Cool idea Eric, I like it.

ericbarefoot committed:
fix: some bugs and style changes

@amoodie pointed out some issues with the PR. changed some importing 
code, and one order of operations problem.

ericbarefoot commented:
Thanks y'all. I will go ahead and merge this when it passes checks.

ericbarefoot committed:
fix: no package import needed for console_scripts!

Surprise! I didn't know how this worked. Turns out you don't need to 
import the whole package anyway. derp derp

amoodie commented:
awesome, thanks @ericbarefoot 

:tada: 10/10 PR experience, would review again :tada:
